channel/quests.go

```diff
@@ -0,0 +1,256 @@
+}
+
+func loadQuestMobKillsFromDB(charID int32) map[int16]map[int32]int32 {
+	out := make(map[int16]map[int32]int32, 16)
+
+	rows, err := common.DB.Query(
+		"SELECT questID, mobID, kills FROM character_quest_kills WHERE characterID=?",
+		charID,
+	)
+	if err != nil {
+		return out
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var qid int16
+		var mobID int32
+		var kills int32
+		if err := rows.Scan(&qid, &mobID, &kills); err != nil {
+			continue
+		}
+		if _, ok := out[qid]; !ok {
+			out[qid] = make(map[int32]int32, 4)
+		}
+		out[qid][mobID] = kills
+	}
+	return out
+}
+
+func upsertQuestRecord(charID int32, questID int16, record string) {
+	_, _ = common.DB.Exec(
+		"INSERT INTO character_quests(characterID, questID, record, completed, completedAt) "+
+			"VALUES(?,?,?,?,?) ON DUPLICATE KEY UPDATE record=VALUES(record), completed=0, completedAt=0",
+		charID, questID, record, 0, 0,
+	)
+}
+
+func setQuestCompleted(charID int32, questID int16, completedAtMs int64) {
+	_, _ = common.DB.Exec(
+		"INSERT INTO character_quests(characterID, questID, record, completed, completedAt) "+
+			"VALUES(?,?,?,?,?) ON DUPLICATE KEY UPDATE completed=1, completedAt=VALUES(completedAt)",
+		charID, questID, "", 1, completedAtMs,
+	)
+}
+
+func deleteQuest(charID int32, questID int16) {
+	_, _ = common.DB.Exec("DELETE FROM character_quests WHERE characterID=? AND questID=?", charID, questID)
+}
+
+func upsertQuestMobKill(charID int32, questID int16, mobID int32, delta int32) {
+	_, _ = common.DB.Exec(
+		"INSERT INTO character_quest_kills(characterID, questID, mobID, kills) VALUES(?,?,?,?) "+
+			"ON DUPLICATE KEY UPDATE kills = kills + VALUES(kills)",
+		charID, questID, mobID, delta,
+	)
+}
+
+func clearQuestMobKills(charID int32, questID int16) {
+	_, _ = common.DB.Exec(
+		"DELETE FROM character_quest_kills WHERE characterID=? AND questID=?",
+		charID, questID,
+	)
+}
+
+// In-memory helpers
+func (q *quests) add(id int16, name string) {
+	q.inProgress[id] = quest{id: id, name: name}
+	delete(q.completed, id)
+
+}
+
+func (q *quests) remove(id int16) {
+	delete(q.inProgress, id)
+}
+func (q *quests) complete(id int16, completedAt int64) {
+	delete(q.inProgress, id)
+	q.completed[id] = quest{id: id, completedAt: completedAt}
+	delete(q.mobKills, id)
+}
+
+func (q *quests) hasInProgress(id int16) bool {
+	_, ok := q.inProgress[id]
+	return ok
+}
+
+func (q *quests) hasCompleted(id int16) bool {
+	_, ok := q.completed[id]
+	return ok
+}
+
+func (q quests) inProgressList() []quest {
+	if len(q.inProgress) == 0 {
+		return nil
+	}
+	out := make([]quest, 0, len(q.inProgress))
+	for _, v := range q.inProgress {
+		out = append(out, v)
+	}
+	return out
+}
+
+func (q quests) completedList() []quest {
+	if len(q.completed) == 0 {
+		return nil
+	}
+	out := make([]quest, 0, len(q.completed))
+	for _, v := range q.completed {
+		out = append(out, v)
+	}
+	return out
+}
+
+// Runtime update packets
+func packetCompleteQuest(questID int16) mpacket.Packet {
```

Comment:
It would be preferable if the function names where `packetQuest....` makes it easier when searching for quest related packets and if a developer is using code completion